testdata/sqllogictests/external_schema.slt

```diff
@@ -1,53 +1,55 @@
-statement ok
-create table src (x int, y text, z int);
-
-statement ok
-insert into src values (1, 'hello', 0), (2, 'other', 10), (3, 'world', 100);
-
-statement ok
-copy (select * from src) to '${TMP}/external_schema.bson'
-
-statement ok
-create external table dst from bson
-options (
-	location '${TMP}/external_schema.bson',
-	file_type 'bson'
-	)
-columns (
-	x int,
-	z int
-);
-
-query ITI
-select * from src;
-----
-1 hello 0 
-2 other 10
-3 world 100
-
-query II
-select * from dst;
-----
-1 0 
-2 10
-3 100
-
-
-statement ok
-create external table dstextra from bson
-options (
-	location '${TMP}/external_schema.bson',
-	file_type 'bson'
-	)
-columns (
-	alpha int,
-	x int,
-	z int
-);
-
-query II
-select * from dstextra;
-----
-NULL 1 0 
-NULL 2 10
-NULL 3 100
+# SKIP FOR NOW
```

Comment:
I guess I don't quite understand what the demarcation between "things that belong in table options" and "things that are an additional option".

Reply:
this has a clear 1-1 mapping to the sql syntax. 

```sql
create external table t 
from <Datasource::NAME> 
[CREDENTIALS] <CredentialOptions> 
[TUNNEL] <TunnelOptions>
[OPTIONS] <TableOptions>
[COLUMNS] <Fields>
```

Reply:
I think that we either need to have: 
- every class of configuration object as a top level object with a clear way to add new kinds of configuration objects, but without an `options` section.
- all configuration classes are components of an `TableOptions` type. 

I do understand that migrating to this might be a multi-step process, but we should agree on what the goal is before getting too much further. 

I also think that a 1-1 mapping with the sql syntax shouldn't be a design goal (and might be a non-goal?)

Reply:
I don't think we should dump everything into `TableOptions`. To me, the `TableOptions` represent the disjunctive union of the create table parameters for that specific discriminant. The intersecting fields are represented in the top level structure.

Reply:
I don't think that's manageable long term, and I don't think it's reasonable to extend the datasource interface every time there's a new "create" parameter: a sizeable percentage of our roadmap are features that exist for some subset of data sources require some configuration, and for which we could choose to either make top a top level object or configuration _or_ embed in options. (access control, dynamic materialized tables, background vacuuming/partitioning operations, etc.) 

We're also aware that we want to change the syntax and a major goal of this refactoring is to make it easier to implement data sources, this solution proposes: that people developing data-sources need to be aware of a legacy create table syntax, and have to change the interface (and every implementation) anytime a "non-disjoint" configuration is  needed, _and_ we're (apparently) opposed to having default implementations of methods that are unused in most datasource implementations... I think we need a design that is a better fit for the needs of the product.

Reply:
> the datasource implementor needs access to all of the possible top level fields (credentials, tunnel, schema, ...)

I think that is not true of all datasource implementers: I don't believe that there is a set of fields or options that are applicable to all data sources. 

- tunnels are only (meaningfully) applicable to data sources that we access over a socket, and currently only implemented for postgres and mysql.
- schemas could be applied anywhere, in theory, though they are mostly applicable to contexts where we do schema inference (json/bson/mongodb/csv): applying them to a relational data source is a thing we have to consider more, and there are some sources (like 
- credentials are only applicable in a set of non-local cases, though credentials are not _always_ used and in many cases sources read credentials out of the `options` block. Also credentials will take a number of different forms depending on the data sources.

Reply:
`DatasourceOptions` seems like it's moving in the right direction, though I don't know that it makes sense to make "tunnels" and "credentials" first class citizens in any of this (and also because credential handling and tunnels are pretty-per datasource as a matter of practice.)